nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeRecord.java

```diff
@@ -323,56 +323,64 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
+            if (flowFiles.isEmpty()) {
+                break;
+            }
+            if (getLogger().isDebugEnabled()) {
+                final List<String> ids = flowFiles.stream().map(ff -> "id=" + ff.getId()).collect(Collectors.toList());
+                getLogger().debug("Pulled {} FlowFiles from queue: {}", ids.size(), ids);
+            }
 
-        final String mergeStrategy = context.getProperty(MERGE_STRATEGY).getValue();
-        final boolean block;
-        if (MERGE_STRATEGY_DEFRAGMENT.getValue().equals(mergeStrategy)) {
-            block = true;
-        } else if (context.getProperty(CORRELATION_ATTRIBUTE_NAME).isSet()) {
-            block = true;
-        } else {
-            block = false;
-        }
+            final String mergeStrategy = context.getProperty(MERGE_STRATEGY).getValue();
+            final boolean block;
+            if (MERGE_STRATEGY_DEFRAGMENT.getValue().equals(mergeStrategy)) {
+                block = true;
+            } else if (context.getProperty(CORRELATION_ATTRIBUTE_NAME).isSet()) {
+                block = true;
+            } else {
+                block = false;
+            }
 
-        try {
-            for (final FlowFile flowFile : flowFiles) {
-                try {
-                    binFlowFile(context, flowFile, session, manager, block);
-                } catch (final Exception e) {
-                    getLogger().error("Failed to bin {} due to {}", new Object[] {flowFile, e});
-                    session.transfer(flowFile, REL_FAILURE);
+            try {
+                for (final FlowFile flowFile : flowFiles) {
+                    try {
+                        binFlowFile(context, flowFile, session, manager, block);
+                    } catch (final Exception e) {
+                        getLogger().error("Failed to bin {} due to {}", flowFile, e);
+                        session.transfer(flowFile, REL_FAILURE);
+                    }
                 }
+            } finally {
+                session.commitAsync();
             }
-        } finally {
-            session.commitAsync();
-        }
 
-        // If there is no more data queued up, or strategy is defragment, complete any bin that meets our minimum threshold
-        // Otherwise, run one more cycle to process queued FlowFiles to add more fragment into available bins.
-        int completedBins = 0;
-        if (flowFiles.isEmpty() || MERGE_STRATEGY_DEFRAGMENT.getValue().equals(mergeStrategy)) {
+            // Complete any bins that have reached their expiration date
             try {
-                completedBins += manager.completeFullEnoughBins();
+                manager.completeExpiredBins();
             } catch (final Exception e) {
-                getLogger().error("Failed to merge FlowFiles to create new bin due to " + e, e);
+                getLogger().error("Failed to merge FlowFiles to create new bin due to {}", e);
```

Comment:
I don't think this change is correct. It'll result in a log message that says `Failed to merge FlowFiles to create new bin due to {}`. Need to either use the `... due to " + e` or `... due to {}", e, e);` so that the first argument matches to the `{}` while the second argument is the `Throwable` argument that causes a stack trace.

Reply:
@markap14 Thanks for catching it. Modified the log statements.